controllers/database_reconciler_helpers.go

```diff
@@ -72,71 +72,72 @@ func (r *DatabaseReconciler) handleDelete(ctx context.Context, database *ndbv1al
+			deregistrationOp, err := ndb_api.DeprovisionDatabase(ctx, ndbClient, database.Status.Id, *ndb_api.GenerateDeprovisionDatabaseRequest())
 			if err != nil {
-				errStatement := "Deprovisioning database instance request failed."
+				errStatement := "Deregistering instance API call failed."
 				log.Error(err, errStatement)
 				r.recorder.Eventf(database, "Warning", EVENT_DEPROVISIONING_FAILED, "Error: %s. %s", errStatement, err.Error())
 				return requeueOnErr(err)
 			}
-		}
-
-		log.Info("Removing Finalizer " + common.FINALIZER_DATABASE_INSTANCE)
-		controllerutil.RemoveFinalizer(database, common.FINALIZER_DATABASE_INSTANCE)
-		if err := r.Update(ctx, database); err != nil {
-			return requeueOnErr(err)
-		}
-		log.Info("Removed Finalizer " + common.FINALIZER_DATABASE_INSTANCE)
-
-	} else if controllerutil.ContainsFinalizer(database, common.FINALIZER_DATABASE_SERVER) {
-		// Checking if the database instance still exists in NDB. (It might take some time for the delete db instance operation to complete)
-		// Proceed to delete the database server vm only after the database instance has been deleted.
-		log.Info("Checking if database instance exists")
-		allDatabases, err := ndb_api.GetAllDatabases(ctx, ndbClient)
-		if err != nil {
-			errStatement := "Error fetching all databases from NDB"
-			log.Error(err, errStatement)
-			r.recorder.Eventf(database, "Warning", EVENT_RESOURCE_LOOKUP_ERROR, "Error: %s. %s", errStatement, err.Error())
-			return requeueOnErr(err)
-		}
-		if len(util.Filter(allDatabases, func(d ndb_api.DatabaseResponse) bool { return d.Id == database.Status.Id })) == 0 {
-			// Could not find the database with the given database id => database instance has been deleted
-			log.Info("Database instance not found, attempting to remove database server.")
-			r.recorder.Eventf(database, "Normal", EVENT_DEPROVISIONING_COMPLETED, "Database deprovisioned from NDB.")
-			r.recorder.Eventf(database, "Normal", EVENT_DEPROVISIONING_STARTED, "Deprovisioning database server from NDB.")
-			databaseServerId := database.Status.DatabaseServerId
-			// Make a dbserver deprovisioning request to NDB only if the serverId is present in status
-			if databaseServerId != "" {
-				_, err := ndb_api.DeprovisionDatabaseServer(ctx, ndbClient, databaseServerId, *ndb_api.GenerateDeprovisionDatabaseServerRequest())
-				if err != nil {
-					errStament := fmt.Sprintf("Deprovisioning database server request failed for id: %s", databaseServerId)
-					log.Error(err, errStament)
-					r.recorder.Eventf(database, "Warning", EVENT_DEPROVISIONING_FAILED, "Error: %s. %s", errStament, err.Error())
+			database.Status.DeregistrationOperationId = deregistrationOp.OperationId
+			if err := r.Status().Update(ctx, database); err != nil {
+				return requeueOnErr(err)
+			}
+		} else {
+			deregistrationOp, err := ndb_api.GetOperationById(ctx, ndbClient, deregistrationOperationId)
+			opStatus := ndb_api.GetOperationStatus(deregistrationOp)
```

Comment:
Sure, I did that because it was being used in both the `elif` cases. Moving them to an uber `else` and eliminating the temp variable by using a switch.